### PR TITLE
Fix disconnect hooks

### LINKED ISSUE #2167: [core] aborting disconnect hook
Hello,

due https://github.com/haraka/Haraka/blob/0c8ae994127b9cd68e2dad1370634c69f2cccc57/plugins.js#L387
it seems that plugin hooks to disconnect event are not executed.

Is it expected behaviour and hook_disconnect should not be really used in production or am I missing something?

```
...
2017-10-02 15:14:01.164201500  [DEBUG] [A1AD977B-E3A7-4F46-84F8-73C3355FA8FD.1] [core] running reset_transaction hook in karma plugin
2017-10-02 15:14:01.165534500  [INFO] [A1AD977B-E3A7-4F46-84F8-73C3355FA8FD.1] [karma] score: -2, good: 0, bad: 0, connections: 59, history: 0, awards: 012,013, fail:rfc5321.MailFrom, rfc5321.RcptTo
2017-10-02 15:14:01.166130500  [DEBUG] [A1AD977B-E3A7-4F46-84F8-73C3355FA8FD.1] [core]  hook=reset_transaction plugin=karma function=hook_reset_transaction params="" retval=CONT msg=""
2017-10-02 15:14:01.176696500  [INFO] [A1AD977B-E3A7-4F46-84F8-73C3355FA8FD.1] [core] client [172.20.0.1] half closed connection
2017-10-02 15:14:01.178321500  [DEBUG] [A1AD977B-E3A7-4F46-84F8-73C3355FA8FD.1] [core] client has disconnected
2017-10-02 15:14:01.178669500  [DEBUG] [A1AD977B-E3A7-4F46-84F8-73C3355FA8FD.1] [core] aborting disconnect hook

```

### system info

Haraka | Haraka.js — Version: 2.8.16
 --- | :--- 
Node | v8.6.0
OS | Linux 3c3555e2d998 4.10.0-041000-generic #201702191831 SMP Sun Feb 19 23:33:19 UTC 2017 x86_64 GNU/Linux
openssl | OpenSSL 1.1.0f  25 May 2017

### PR DESCRIPTION
**Pull request by smfreegard**
Fixes #2167

This is a really strange one - something changed in core between 2.8.14 and 2.8.16 that broke the handling of disconnects but I'm not sure exactly what (as the code here hasn't changed for 2 years).

Currently master is so broken, that if you send a QUIT - it doesn't actually end the socket and hook_disconnect is *never* run, so this is a pretty serious bug.

I've had this running on my new trap for a few hours without issue.

### CODE REVIEW

File: plugins.js [hunk 1]
```diff
@@ -384,20 +384,18 @@ plugins._register_plugin = function (plugin) {
 };
 
 plugins.run_hooks = function (hook, object, params) {
-    if (client_disconnected(object)) {
-        if (hook !== 'log') {
-            object.logdebug('aborting ' + hook + ' hook');
-        }
+    if (client_disconnected(object) && !is_required_hook(hook)) {
+        object.logdebug('aborting ' + hook + ' hook');
```
Comment (msimerson) [inline plugins.js]: this would be better as a string literal:

```js
object.logdebug(`aborting hook: ${hook}`);
```

**Comment by smfreegard** [comment]
I'm not changing stuff like that - I'll leave that to the #hacktoberfest lot to do the whole file in one go.

**Comment by analogic** [comment]
Thanks, it works! But there seems still be issue with disconnecting socket and minor issue repeating "client has disconnected" by every hook 

```
...
2017-10-05 06:20:43.755326500  [PROTOCOL] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] C: QUIT state=1
2017-10-05 06:20:43.755439500  [DEBUG] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] running quit hooks
2017-10-05 06:20:43.755627500  [PROTOCOL] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] S: 221 ... closing connection. Have a jolly good day.
2017-10-05 06:20:43.755876500  [DEBUG] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] client has disconnected
2017-10-05 06:20:43.755960500  [DEBUG] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] running disconnect hooks
2017-10-05 06:20:43.756039500  [DEBUG] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] client has disconnected
2017-10-05 06:20:43.756149500  [DEBUG] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] running disconnect hook in log_transaction plugin
2017-10-05 06:20:43.756756500  [DEBUG] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] client has disconnected
2017-10-05 06:20:43.756876500  [DEBUG] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core]  hook=disconnect plugin=log_transaction function=log_connection_end params="" retval=CONT msg=""
2017-10-05 06:20:43.756915500  [DEBUG] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] client has disconnected
2017-10-05 06:20:43.757062500  [DEBUG] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] running disconnect hook in tls plugin
2017-10-05 06:20:43.757167500  [DEBUG] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] client has disconnected
2017-10-05 06:20:43.757277500  [DEBUG] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core]  hook=disconnect plugin=tls function=hook_disconnect params="" retval=CONT msg=""
2017-10-05 06:20:43.757447500  [NOTICE] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] disconnect ip=172.20.0.1 rdns=NXDOMAIN helo=foo.com relay=N early=N esmtp=N tls=N pipe=N errors=0 txns=1 rcpts=1/0/0 msgs=1/0/0 bytes=87 lr="" time=0.105
2017-10-05 06:20:43.757583500  [PROTOCOL] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] C: (after-disconnect): QUIT state=100
2017-10-05 06:20:43.757751500  [INFO] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] data after disconnect from 172.20.0.1
2017-10-05 06:20:43.757838500  [PROTOCOL] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] C: (after-disconnect): QUIT state=100
2017-10-05 06:20:43.757945500  [INFO] [CE3EABE9-4D91-4B8F-817A-D7F3C435C333.1] [core] data after disconnect from 172.20.0.1
```

**Comment by smfreegard** [comment]
>  But there seems still be issue with disconnecting socket and minor issue repeating "client has disconnected" by every hook

That's a DEBUG level entry, so I'm not really concerned about it.

**Pull request merged**